pkg/server/auth.go

```diff
@@ -161,3 +165,218 @@ func (s *Server) Logout(c *gin.Context) {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Set cookies for the client
+	if err = auth.SetAuthCookies(c, out.AccessToken, out.RefreshToken); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Content negotiation and redirection if required.
+	switch c.NegotiateFormat(binding.MIMEJSON, binding.MIMEHTML) {
+	case binding.MIMEJSON:
+		c.JSON(http.StatusOK, out)
+	case binding.MIMEHTML:
+		htmx.Redirect(c, http.StatusSeeOther, in.Redirect())
+	default:
+		c.AbortWithError(http.StatusNotAcceptable, errors.ErrNotAccepted)
+	}
+}
+
+// Reauthenticate a user via their refresh token.
+func (s *Server) Reauthenticate(c *gin.Context) {
+	var (
+		err    error
+		claims *gimlet.Claims
+		sub    gimlet.SubjectType
+		subID  ulid.ULID
+		in     *api.ReauthenticateRequest
+		out    *api.LoginReply
+	)
+
+	if err = c.BindJSON(&in); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusBadRequest, api.Error(errors.ErrBindJSON))
+		return
+	}
+
+	// Validate the request
+	if err = in.Validate(); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusBadRequest, api.Error(err))
+		return
+	}
+
+	// Verify the refresh token
+	if claims, err = s.issuer.Verify(in.RefreshToken); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusForbidden, api.Error(errors.ErrFailedAuthentication))
+		return
+	}
+
+	// TODO: Verify that this is a refresh token with the correct audience.
+
+	// Parse the subject type and ID from the claims.
+	if sub, subID, err = claims.SubjectID(); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Load claims based on the subject type.
+	switch sub {
+	case gimlet.SubjectUser:
+		if claims, err = s.reauthenticateUser(c, subID); err != nil {
+			// Error logging is handled in reauthenticateUser
+			return
+		}
+	case gimlet.SubjectAPIKey:
+		if claims, err = s.reauthenticateAPIKey(c, subID); err != nil {
+			// Error logging is handled in reauthenticateAPIKey
+			return
+		}
+	default:
+		c.Error(fmt.Errorf("unknown subject type: %c", sub))
+		c.JSON(http.StatusForbidden, api.Error(errors.ErrFailedAuthentication))
+		return
+	}
+
+	// Create new access and refresh tokens
+	out = &api.LoginReply{}
+	if out.AccessToken, out.RefreshToken, err = s.issuer.CreateTokens(claims); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Set cookies for the client
+	if err = auth.SetAuthCookies(c, out.AccessToken, out.RefreshToken); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Content negotiation and redirection if required.
+	switch c.NegotiateFormat(binding.MIMEJSON, binding.MIMEHTML) {
+	case binding.MIMEJSON:
+		c.JSON(http.StatusOK, out)
+	case binding.MIMEHTML:
+		htmx.Redirect(c, http.StatusSeeOther, in.Redirect())
```

Comment:
Let's remove the `in.Redirect()` methods -- see how we're doing it in `Login` now to use the configured redirect link.